File: attempted/p152/p152.scala
import scala.collection.mutable

object P152 {
  val primesAndMaxPowers =
    Seq(
      2 -> 6,
      3 -> 4,
      5 -> 2,
      7 -> 2,
      11 -> 1,
      13 -> 1,
      17 -> 1,
      19 -> 1,
      23 -> 1,
      29 -> 1,
      31 -> 1,
      37 -> 1)

  private def buildExponents(maxExponents: Seq[Int]): Seq[List[Int]] = {
    val items = mutable.Buffer.empty[List[Int]]

    def build(maxes: List[Int], current: List[Int]) {
      if(maxes.isEmpty) {
        items += current
        return
      }

      val nextMax::remaining = maxes

      var e = 0
      while(e <= nextMax) {
        build(remaining, e :: current)
        e += 1
      }
    }

    build(maxExponents.reverse.toList, List())

    items.toSeq
  }

  def generateGCD = buildExponents(primesAndMaxPowers.unzip._2)

  val all1 = (1l << primesAndMaxPowers.length) - 1
  def pack(used: Seq[Int]) = {
    var result = 0l
    var index = 0
    used.foreach {
      case u if u > 0 => 
        result |= (1l << index)
        index += 1
      case _ =>
        index += 1
    }

    result.toLong
  }

  def unpack(used: Long) = {
    (0 until primesAndMaxPowers.length).map { ind =>
      ((used >> ind) & 1) == 1
    }
  }

  def countCoprime(gcd: Seq[Int]): Int = {
    //println(s"gcd exponents: ${gcd}")
    //println(s"primes used by gcd: ${unpack(pack(gcd))}")
    def buildResiduals(used: Long): Seq[(Long, List[Int])] = {
      val maxPowers = primesAndMaxPowers.unzip._2.zipWithIndex.map {
        case (exp, ind) if ((used >> ind) & 1l) == 0 => 
          exp
        case (exp, _) =>
          0
      }

      val residualsFactored = buildExponents(maxPowers)
      residualsFactored.map(rf => (used | pack(rf), rf))
    }

    import mutable.Queue

    val q = Queue.empty[(Long, Set[List[Int]])]

    q.enqueue((pack(gcd), Set()))
    
    var count = 0

    while(!q.isEmpty) {
      val (used, residuals) = q.dequeue

      val nextResiduals = buildResiduals(used)
      //println(s"got: ${residuals}, begets ${nextResiduals.length} new ones")
      nextResiduals.foreach {
        case (nextUsed, _) if nextUsed == all1 =>
        case (_, newResidual) if residuals.contains(newResidual) =>
        case (nextUsed, newResidual) =>
          q.enqueue((nextUsed, residuals + newResidual))
          count += 1
      }
    }

    count
  }
}

val gcds = P152.generateGCD
println(s"# GCDs: ${gcds.length}")

println(s"packed(gcd(20000)) == ${P152.pack(gcds(20000))}")

gcds.zipWithIndex.foreach { case (gcd, ind) =>
  println(s"count_coprime(gcd($ind)) == ${P152.countCoprime(gcd)}")
}

File: attempted/p152/p152.py
#!/usr/bin/python

valid_primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37]
max_powers =   [6, 4, 2, 2,  1,  1,  1,  1,  1,  1,  1,  1]

def build(max_exponents):
    results = []
    def build(ex):
        if len(ex) == len(valid_primes):
            results.append(ex)
            return

        ind = len(ex)
        for e in range(1 + max_exponents[ind]):
            build(ex + [e])
        
    build([])

    return results

def generate_gcd():
    return build(max_powers)

def count_coprime(gcd):
    from Queue import Queue

    def build_residuals(primes_used):
        max_powers_lcl = [0 if used else e for (e, used) in zip(max_powers, primes_used)]

        def merge_used(ex):
            return [(ex[i] | primes_used[i]) for i in range(len(ex))]

        residuals_factored = build(max_powers_lcl)

        return [(merge_used(ex), None) for ex in residuals_factored]

    primes_used_by_gcd = [e > 0 for e in gcd]
    q = []
    q.append((primes_used_by_gcd, set())) # store used_primes bitfield and list of residuals

    print('primes used by gcd: %s' % str(primes_used_by_gcd))

    print('num residuals: %d' % len(build_residuals(primes_used_by_gcd)))

    count = 0
    while len(q) > 0:
        (primes_used, current_residuals) = q[0]
        q = q[1:]


        next_residuals = build_residuals(primes_used)
        print('got: %s, begets %d new ones' % (','.join(map(str, current_residuals)), len(next_residuals)))
        for (new_primes_used, next_residual) in next_residuals:
            done = all(new_primes_used)

            if done:
                continue

            if next_residual in current_residuals:
                # should only happen for 1
                continue

            new_residuals = current_residuals | set([next_residual])
            q.append((new_primes_used, new_residuals))
            count += 1

    return count

gcds = generate_gcd()

print len(gcds)

print('gcds[20000]: %s' % str(gcds[20000]))

print count_coprime(gcds[20000])
